compile on MinGW/MSYS2

dsogari opened the pull request:
Hello,
I'm reporting to you the problems I encountered when trying to compile libkml from your repo in MinGW/MSYS2, and how I managed to fix or workaround them:
- add `strptime.c` from NetBSD, which provides the `strptime` function
- add gitattributes to testdata folder, for checking out text files with Unix line endings
- patch installation of CMake and pkgconfig configuration files for MinGW
- patch `build_target` macro to link libraries without the `LINK_INTERFACE_LIBRARIES` specifier (due to reasons unknown to me, it was giving undefined references when linking kmldom)
- patch `super_find_package` macro to propagate some result variables to the parent scope (needed for zlib, strangely)
- link `curlfetch` from the hellonet example as static library and exchange the link order of the tests
- apply similar patch for the linking of the helloworld example
- remove inclusion of `termios.h` from several source files in the hellonet example (`prompt.h` already provides what is necessary)
- implement disabling/enabling of echo in standard input on Windows in the hellonet example (see `prompt.h`)
- set properties on some source files to suppress specific warnings
- define symbol in `unzip.c` to suppress warnings about different data types in function calls
- use the correct implementation of `strptime` function in `date_time.cc`
- fix the use of wide strings in `file_win32.cc` (was not explicitly necessary)
- define `strncasecmp` in `string_util.cc` only if it was not previously defined
- patch the `FilenameToUri` function in `uri_parser.cc` to depend on the file path itself instead of on the platform
- ad hoc implementation of `getenv` in `google_doc_list.cc` was not needed
- specify the minimum required version of Python, so as to choose Python 2, because Python 3 complained about the syntax of the command used to obtain the installation directory
- fix initialization of static variable in `Testb2a_hex` from `string_util_test.cc`
- fix the group name of the convenience tests in the call to `build_test` (was regionator, which caused some weird errors about overwriting object files)
- provide `setenv` function in `google_maps_data_test.cc`
- select the correct expected number on `TestComputeBboxLookAt` depending on architecture
- select the correct path strings on some tests of `KmzTest`, to account for the path separator on each platform
- [updated] add 64-bit specific compilation flag for the python bindings on windows (it was giving undefined reference to `_imp __Py_InitModule4` on appveyor in the MSYS2 project)

dsogari committed:
compile and test libkml on MinGW from the MSYS2 project:
 - add strptime.c from NetBSD, which provides the strptime function
 - add gitattributes to testdata folder, for checking out text files with Unix line endings
 - patch installation of CMake and pkgconfig configuration files for MinGW
 - patch build_target macro to link libraries without the LINK_INTERFACE_LIBRARIES specifier (due to reasons unknown to me)
 - patch super_find_package macro to propagate some result variables to the parent scope (needed for zlib, strangely)
 - link curlfetch from the hellonet example as static library and exchange the link order for the tests
 - apply similar patch for the linking of the helloworld example
 - remove inclusion of termios.h from several source files in the hellonet example (prompt.h already provides what is necessary)
 - implement disabling/enabling of echo in standard input on Windows in the hellonet example (see prompt.h)
 - set properties on some source files to suppress specific warnings
 - define symbol in unzip.c to suppress warnings about different data types in function calls
 - use the correct implementation of strptime function in date_time.cc
 - fix the use of wide strings in file_win32.cc (was not explicitly necessary)
 - define strncasecmp in string_util.cc only if it was not previously defined
 - patch the FilenameToUri function in uri_parser.cc to depend on the file path itself instead of on the platform
 - ad hoc implementation of getenv in google_doc_list.cc was not needed
 - specify the minimum required version of Python, so as to choose Python 2, because Python 3 complained about the syntax of the command used to obtain the installation directory
 - fix initialization of static variable in Testb2a_hex from string_util_test.cc
 - fix the group name of the convenience tests in the call to build_test (was regionator, which caused some weird errors about overwriting object files)
 - provide setenv function in google_maps_data_test.cc
 - select the correct expected number on TestComputeBboxLookAt depending on architecture
 - select the correct path strings on some tests of KmzTest, to account for the path separator on each platform

rashadkm commented:
Thanks for this PR for support on mingw. The bulletin list really was helpful when looking at this code. I really appreciate your effort. I would like to find the problem in python2 vs python3 about installation directory and see if we could support both or python3.

Everything else looks good and ready to merge.

rashadkm commented on tests/kml/base/CMakeLists.txt:
```diff
@@ -22,6 +22,11 @@ uri_parser
 
+if(MINGW)
+  #supress overflow warning for mingw
+  set_source_files_properties(${CMAKE_CURRENT_SOURCE_DIR}/expat_parser_test.cc COMPILE_FLAGS "-Wno-overflow")
```
is it really an overflow that could be ignored ?

rashadkm commented on tests/kml/engine/kmz_file_test.cc:
```diff
@@ -370,7 +374,11 @@ TEST_F(KmzTest, TestCreateFromElement) {
   ASSERT_EQ(static_cast<size_t>(3), list.size());
   ASSERT_EQ(string("doc.kml"), list[0]);
   ASSERT_EQ(string("dummy.png"), list[1]);
+#ifndef __MINGW32__
```
should this be _WIN32. because it should be same for msvc right?

rashadkm commented on tests/kml/engine/kmz_file_test.cc:
```diff
@@ -431,10 +447,17 @@ TEST_F(KmzTest, TestCreateFromGoogleEarthFile) {
+  ASSERT_EQ(string("files\\camelbrown200.png"), list[1]);
+  ASSERT_EQ(string("files\\camelblack200.png"), list[2]);
+  ASSERT_EQ(string("files\\camera_mode.png"), list[3]);
+  ASSERT_EQ(string("files\\camelcolor200.png"), list[4]);
```
same comment as above

dsogari committed:
follow-up:
 - revert changes in file_win32.cc (define UNICODE in CMakeLists.txt instead)
 - remove const qualifier from return type in simplifylines.cc
 - do not suppress overflow warning in expat_parser_test.cc (define XML_UNICODE in CMakeLists.txt instead)
 - use _WIN32 instead of __MINGW32__ in conditional compilation of kmz_file_test.cc
 - remove Python 2 requirement (change print command to the Python 3 syntax instead)

ghost commented:
@rashadkm thanks for the feedback! I already made some adjustments to reflect your queries. I also reverted the changes in `file_win32.cc`, because I thought best to keep the Unicode support.

Regarding Python 3, I had to add a compile option to suppress warnings about "missing-field-initializers". However, I could not make the tests work with Python 3. They say `No module named 'kmlbase'` or `No module named 'kmldom'`, and I did not know how to solve that :(

rashadkm commented:
Thanks.
